src/components/abstractComponent.ts

```diff
@@ -0,0 +1,436 @@
+/**
+ * Copyright 2014-present Palantir Technologies
+ * @license MIT
+ */
+
+import * as d3 from "d3";
+
+import { ComponentContainer } from "./componentContainer";
+
+import { Point, SpaceRequest, Bounds } from "../core/interfaces";
+import * as RenderController from "../core/renderController";
+import * as Utils from "../utils";
+import { Set } from "../utils/set";
+
+export type GenericComponentCallback<D> = (component: IComponent<D>) => void;
+
+export type IResizeHandler = (size: { height: number, width: number }) => void;
+
+export interface IComponent<D> {
```

Comment:
Also please add a comment to this indicating what this template type is. Is it the type of the container? HTMLElement vs SVGElement? Various instances of this type elsewhere in this PR simply have `IComponent<any>` so it's hard to tell what this is for.